Feature abfe somd

fjclark opened the pull request:
Hello,

I've begun work implementing ABFE for SOMD in BSS. This is not yet ready to be merged, but I have opened the pull request to allow discussion of the code. At the moment, the following code seems to work:

```
import  BioSimSpace.Sandpit.Exscientia as BSS
from BioSimSpace.Sandpit.Exscientia.Align import decouple

# Load system (already equilibrated) and mark ligand as decoupled
system = BSS.IO.readMolecules(["input/SYSTEM.crd", "input/SYSTEM.top"])
lig = decouple(system[0])
system.updateMolecule(0,lig)

# Find the restraints
protocol = BSS.Protocol.Production(timestep=2*BSS.Units.Time.femtosecond, runtime=0.01*BSS.Units.Time.nanosecond)
restraint_search = BSS.FreeEnergy.RestraintSearch(system, protocol=protocol, engine='SOMD', gpu_support=True, work_dir='restraint_search')
restraint_search.start()
restraint = restraint_search.analyse(block = True)

# Carry out the restraining stage 
restrain_protocol = BSS.Protocol.FreeEnergy(lam_vals=[0.0, 0.25, 0.5, 1], perturbation_type="restraint")
restrain_fe_calc = BSS.FreeEnergy.Relative(restraint.system, restrain_protocol, engine='somd', restraint=restraint, work_dir='restrain')
restraint_fe_calc.run()
```
But I am yet to implement a full set of tests. Note that restraint_search.analyse() will fail for versions of Sire older than yesterday due a bug in the code used to write the dcd files.

Some points for discussion:

- Currently, my code completely breaks ABFE for GROMACS because I have changed the use of property maps in _decouple.py. I will fix this. However, my understanding is that the use of property maps as originally implemented here (where they are used to specify perturbations) is inconsistent with the rest of BSS (where they are used to allow user-specified property names). Could this be changed to maintain consistency throughout BSS?

- Could restraint.correction be implemented as a method, rather than a property? I intend to implement the semi-analytical correction for Boresch restraints in the future, and this could allow users to select the units and the method, e.g. `restraint.correction(method ="numerical", units=kj_per_mol)`.

- As shown above, the current code only allows for one stage of the FE calc to be completed in SOMD at a time. The complete calculation would require BSS.FreeEnergy.Relative to be run three times with three different protocols, with perturbation types "restraint", "discharge_soft", and "vanish_soft". I wonder if the best way round this would be to allow passing of a list of protocols to BSS.FreeEnergy.Relative?

I am away for the next two weeks, but will resume work on this in August, when I will:

- Tidy this up and write tests
- Work on the restraint derivation code. We would like to make some additions to what is available in MDRestraintsGenerator. It seems that the best way to do this may be to copy the code directly into BSS and modify it there. 

Any comments would be appreciated. 

Thanks,
Finlay

xiki-tempula commented:
> Currently, my code completely breaks ABFE for GROMACS because I have changed the use of property maps in _decouple.py. I will fix this. However, my understanding is that the use of property maps as originally implemented here (where they are used to specify perturbations) is inconsistent with the rest of BSS (where they are used to allow user-specified property names). Could this be changed to maintain consistency throughout BSS?

Yes, I need to change this.

> Could restraint.correction be implemented as a method, rather than a property? I intend to implement the semi-analytical correction for Boresch restraints in the future, and this could allow users to select the units and the method, e.g. restraint.correction(method ="numerical", units=kj_per_mol).

I think both of them are fine? The you could have both `restraint.correction` and `restraint.correction()`

> As shown above, the current code only allows for one stage of the FE calc to be completed in SOMD at a time. The complete calculation would require BSS.FreeEnergy.Relative to be run three times with three different protocols, with perturbation types "restraint", "discharge_soft", and "vanish_soft". I wonder if the best way round this would be to allow passing of a list of protocols to BSS.FreeEnergy.Relative?

I think you could just do this within the protocol?
```
restraint_protocol = FreeEnergyMinimisation(
            lam=pd.Series(data={'bonded': 0.0, }),
            lam_vals=pd.DataFrame(
                data={'bonded': [0.0, 0.25, 0.5, 0.75, 1.0, 1.0, 1.0]}),
            steps=10000)

restraint_protocol = FreeEnergyMinimisation(
            lam=pd.Series(data={'coul': 0.0, }),
            lam_vals=pd.DataFrame(
                data={'coul': [0.0, 0.25, 0.5, 0.75, 1.0, 1.0, 1.0]}),
            steps=10000)

restraint_protocol = FreeEnergyMinimisation(
            lam=pd.Series(data={'vdw': 0.0, }),
            lam_vals=pd.DataFrame(
                data={'vdw': [0.0, 0.25, 0.5, 0.75, 1.0, 1.0, 1.0]}),
            steps=10000)
```

xiki-tempula commented:
> Yes, but I was wondering if it would be cleaner to run the entire bound or free leg at once by passing several protocols to a single BSS.FreeEnergy.Relative. This is probably fine as it is though.

I'm happy either way as in Gromacs one only require one protocol. If Lester is happy, I don't really care.

xiki-tempula commented:
> Currently, my code completely breaks ABFE for GROMACS because I have changed the use of property maps in _decouple.py. I will fix this. However, my understanding is that the use of property maps as originally implemented here (where they are used to specify perturbations) is inconsistent with the rest of BSS (where they are used to allow user-specified property names). Could this be changed to maintain consistency throughout BSS?

I have adopted Lester's suggestions and you could see the new code in the dev branch now.

fjclark commented:
Hi @xiki-tempula, I've now made most of the changes I was planning - ABFE works for SOMD and I've tidied up and improved my code for generating Boresch restraints. 

I've implemented BSS.FreeEnergy.Absolute as a derived class of BSS.FreeEnergy.Relative - there is almost no difference other than Absolute requires that restraints are passed, and Relative does not accept restraints. This is purely for clarity. Please let me know what you think.

I've been having some issues with the tests - for example, the tests at test/Sandpit/Exscientia/Free_Energy/ occasionally all fail due to inability to find an input file:
``` 
ERROR test/Sandpit/Exscientia/Free_Energy/test_restraint.py::TestSomdOutput::test_equil_vals - OSError: Missing input file: 'test/input/ligands/ligand01.prm7'
```
Other tests which I shouldn't have effected with my work (e.g. in test_merge.py) also seem to have similar problems. The tests seem to be more reliable when I test each module individually. Do you have any suggestions as to what the issue might be? Although the C.I. is failing for my latest push, this is only due to two errors on tests which consistently run fine locally, and occasionally work through the C.I.

Could you please review my code when you have the time?

Thanks very much.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_relative.py:
```diff
@@ -1395,7 +1387,10 @@ def _initialise_runner(self, system):
         # Name the first directory.
-        first_dir = f"{self._work_dir}/lambda_{self._protocol.getLambdaIndex()}"
+        if self._engine == "SOMD":
+            first_dir = f"{self._work_dir}/lambda_{lam.values[0]:.3f}"
```
Is there any reason why we cannot do `lambda_{self._protocol.getLambdaIndex()}`?

fjclark replied:
As SOMD does not allow for multi-lambda protocols, I thought it was more informative to make the value of lambda obvious by keeping the previous naming scheme. However, happy to change this for consistency if you think it would be better.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint.py:
```diff
@@ -339,5 +397,14 @@ def correction(self):
+            if unit == "kj_per_mol":
+                dG *= kj_per_mol
+            elif unit == "kcal_per_mol":
+                dG *= (kj_per_mol/kcal_per_mol)*kcal_per_mol
```
why do you need to do this? If there is unit attached to it then there is no need specify it?
if you want kcal_per_mol, you could do
`energy = dG / kcal_per_mol`
if you want kj_per_mol, you could do
`energy = dG / kj_per_mol`

fjclark replied:
Good point - I wasn't thinking. Thanks for pointing this out.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -270,7 +272,7 @@ def kill(self):
     def isRunning(self):
         """Check if the process is running."""
-        self._process.isRunning()
+        return self._process.isRunning()
```
thanks for catching this

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -288,50 +290,55 @@ def _analyse(self, rest_type='Boresch',
+                Default is 'Boresch'.
+
+            method: str
+                The method to use to derive the restraints. Currently only 'MDRestraintsGenerator'
```
Do you want to change this as well?

fjclark replied:
Thanks for catching this.

xiki-tempula commented on python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py:
```diff
@@ -405,7 +413,8 @@ def analyse(work_dir, system, traj, temperature, rest_type='Boresch',
                 recept_selection_str='protein and name CA C N',
-                cutoff=10): # In Angstrom
+                cutoff=10,
+                restraint_idx=0): # In Angstrom
```
wrong place for comment

fjclark replied:
Thanks

xiki-tempula commented on test/Sandpit/Exscientia/Align/test_decouple.py:
```diff
@@ -61,3 +64,18 @@ def test_topology(mol, tmp_path):
+        assert atom.property("charge1") == 0*_SireUnits.e_charge
+        assert atom.property("LJ1") == _SireMM.LJParameter()
+        assert atom.property("element1") == _SireMol.Element(0)
+        assert atom.property("ambertype1") == "du"
```
extra line

xiki-tempula commented on test/Sandpit/Exscientia/Free_Energy/test_absolute.py:
```diff
@@ -0,0 +1,155 @@
+    @staticmethod
+    @pytest.fixture(scope='class')
+    def freenrg():
+        m = BSS.Parameters.openff_unconstrained_2_0_0(
```
might be good to have pre-parametrised ligand like m = BSS.IO.readMolecules("test/input/amber/ala/*").getMolecule(0)?

fjclark replied:
Definitely, thanks.

xiki-tempula commented on test/Sandpit/Exscientia/Free_Energy/test_restraint_search.py:
```diff
@@ -130,3 +145,72 @@ def test_index(self, restraint_search):
+        idxs = {k:restraint._restraint_dict['anchor_points'][k].index() for k \
+                in restraint._restraint_dict['anchor_points']}
+        assert idxs == {'r1':1560, 'r2':1558, 'r3':1562,
+                        'l1':9, 'l2':8, 'l3':10}
```
extra line

xiki-tempula commented:
@fjclark Are you in a rush to have this merged? I think it would be the best if @lohedges could review this.

fjclark commented:
@xiki-tempula thanks very much for the comments - I'll address these soon. 

No, there is no rush. We're giving a workshop on BSS on the 20th of September where we plan to have a short section on ABFE in BSS (as demonstration of the sandpit/ experimental features). However, we'll just create a temporary `feature-workshop` branch in the main repo, and I'll pull my branch directly into that.

xiki-tempula commented:
@fjclark Great, I would imagine that @lohedges would demand some large changes so might be good for me to review it after.

fjclark commented:
Hi @lohedges, could you please review my implementation of ABFE for SOMD when you have time? As large changes are likely to be required, @xiki-tempula would prefer to check after you've had a look. My main changes to the Exscientia main branch are:

- Implementing ABFE for SOMD following @xiki-tempula's implementation for GROMACS
-  Creating an `Absolute` class derived from `Relative`. This is almost identical to `Relative`, but it accepts restraints and I've modified `Relative` so that it does not. This is purely for clarity, so that Absolute calculations are not run using `Relative`. However, the way I've implemented this is likely poor (e.g. (this)[https://github.com/fjclark/BioSimSpace/blob/27c09b48b4099bfbf27fcb421afb5c30a6a42e12/python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_absolute.py#L141]). Any suggestions for a better implementation would be brilliant.
- Creating a native BSS method for selecting anchor atoms for Boresch restraints
- Modifying the analytical correction for Boresch restraints in _restraint.py to work when some of the force constants are 0, and adding the more reliable semi-analytical correction (mainly from YANK).

@xiki-tempula, I've modified the code to address your initial comments. Please let me know if you have any more feedback.

Thanks

lohedges commented:
Hi @fjclark. Apologies for missing this. I'll put it on my to-do list for next week.
